### PR TITLE
add canonical flag for DNSSEC signature

### PR DESCRIPTION
**Pull request by khibino**
Normalized RR encoders are required when calculating the value of the RRSIG signature

### CODE REVIEW

File: dnsext-types/DNS/Types/Domain.hs [hunk 1]
```diff
@@ -182,25 +183,38 @@ modifyMailbox f (Mailbox o l) = Mailbox (f o) (f l)
...
 ----------------------------------------------------------------
 
-putDomain :: Domain -> SPut
-putDomain (Domain o _) = putDomain' _period o
+-- | Canonical-form flag.
+-- For example, `Canonical` is used from DNSSEC extension.
+--
+-- ref. https://datatracker.ietf.org/doc/html/rfc4034#section-6.2 - Canonical RR Form
+data CanonicalFlag
+  = NoCanonical
```
Comment (kazu-yamamoto) [inline dnsext-types/DNS/Types/Domain.hs:191]: `Compression` would be better than `NoCanonical`.

File: dnsext-types/DNS/Types/Domain.hs [hunk 2]
```diff
@@ -182,25 +183,38 @@ modifyMailbox f (Mailbox o l) = Mailbox (f o) (f l)
...
 
-putMailbox :: Mailbox -> SPut
-putMailbox (Mailbox o _) = putDomain' _at o
+putMailbox :: CanonicalFlag -> Mailbox -> SPut
+putMailbox NoCanonical (Mailbox o _) = putDomain' _at True  o
+putMailbox Canonical   (Mailbox _ l) = putDomain' _at False l {- canonical form is lowercase and no name-compression. -}
 
-putDomain' :: Word8 -> RawDomain -> SPut
-putDomain' sep dom
+putDomain' :: Word8 -> Bool -> RawDomain -> SPut
```
Comment (kazu-yamamoto) [inline dnsext-types/DNS/Types/Domain.hs:205]: I'm not sure why `Bool` is used here. Why don't you use `CanonicalFlag` directly?

**Review by kazu-yamamoto — CHANGES_REQUESTED** [review]
I think you can improve the naming and the signatures of functions.

File: dnsext-types/DNS/Types/Domain.hs (hunk 1 above)
Comment (khibino) [reply dnsext-types/DNS/Types/Domain.hs]: https://github.com/kazu-yamamoto/dnsext/pull/13/commits/3bfc0d3c9b6a2ac5623f0ccf74b5eb8ad435594d

File: dnsext-types/DNS/Types/Domain.hs (hunk 2 above)
Comment (khibino) [reply dnsext-types/DNS/Types/Domain.hs]: https://github.com/kazu-yamamoto/dnsext/pull/13/commits/faf2131a34d207ce852fd169a190f2b4ca06a151

File: dnsext-types/DNS/Types/Domain.hs [hunk 3]
```diff
@@ -182,25 +183,38 @@ modifyMailbox f (Mailbox o l) = Mailbox (f o) (f l)
...
+putDomain' :: Word8 -> CanonicalFlag -> RawDomain -> SPut
+putDomain' sep cf dom
     | Short.null dom || dom == rootDomain = put8 0
     | otherwise = do
         mpos <- popPointer dom
         cur <- builderPosition
         case mpos of
-            Just pos -> putPointer pos
-            Nothing  -> do
+            Just pos | Compression <- cf  -> putPointer pos
```
Comment (kazu-yamamoto) [inline dnsext-types/DNS/Types/Domain.hs:212]: Use `==` with Eq instance added.

**Review by kazu-yamamoto — CHANGES_REQUESTED** [review]
I like orthodox programming style.

File: dnsext-types/DNS/Types/Domain.hs (hunk 3 above)
Comment (khibino) [reply dnsext-types/DNS/Types/Domain.hs]: https://github.com/kazu-yamamoto/dnsext/pull/13/commits/d2647519f6279c578cf610d72424cbe9ba59979b

**Review by kazu-yamamoto — APPROVED** [review]
LGTM.

**Comment by kazu-yamamoto** [comment]
Merged.

**Pull request merged**